src/python/pants/bsp/protocol.py

```diff
@@ -124,6 +128,15 @@ def _handle_inbound_message(self, *, method_name: str, params: Any):
                 )
             )
 
+        # Handle the build/shutdown method and build/exit notification.
+        if method_name == self._SHUTDOWN_METHOD_NAME:
+            return None
+        elif method_name == self._EXIT_NOTIFCATION_NAME:
+            # TODO: Figure how to exit with a success error code. `sys.exit(0)` results in:
+            #   `Signal 15 (SIGTERM) was raised. Exiting with failure.`
+            os.kill(os.getpid(), signal.SIGKILL)
```

Comment:
`sys.exit(0)` made the process exit with failure whereas the BSP spec requires it exit cleanly. Is there a better way to do this? 

I might have to vendor the LSP/BSP protocol impl we are using to add support for exiting out of the processing loop. Pants should exit cleanly on SIGTERM though (which sys.exit(0) triggered???).

Reply:
Hm. This one is questionable.

If the goal is to tear down `pantsd`, I'd suggest adding a lifecycle hook injected from `DaemonPantsRunner` (which starts the `LocalPantsRunner`, which starts the `BuiltinGoal`) here https://github.com/pantsbuild/pants/blob/0246426357458118db75aae7260ed0b863e20aab/src/python/pants/bin/daemon_pants_runner.py#L131-L137

The `DaemonPantsRunner` holds a reference to the `PantsDaemonCore`, which has a kill switch you could expose: https://github.com/pantsbuild/pants/blob/0246426357458118db75aae7260ed0b863e20aab/src/python/pants/pantsd/pants_daemon_core.py#L55-L56 ... when set, the server will exit cleanly.

Reply:
Interesting, so the read-dispatch loop wouldn't need to exit as well in order to return to whatever spawns the session?

Reply:
> This one is questionable.

Exactly! Itwasn't clear to me what the right solution here was. Will investigate the suggested "kill switch".

Reply:
> Interesting, so the read-dispatch loop wouldn't need to exit as well in order to return to whatever spawns the session?

Well, yes: this loop would need to exit, as that is what would cause the client to disconnect, I think. The client needs to disconnect because:

> ... when set, the server will exit cleanly after the client disconnects.

Reply:
Makes sense. Points at the need then to vendor the read-dispatch loop we use from the `python-lsp-jsonrpc` package since it doesn't include a way to force the loop to exit.

Reply:
> Makes sense. Points at the need then to vendor the read-dispatch loop we use from the `python-lsp-jsonrpc` package since it doesn't include a way to force the loop to exit.

Assuming that `SHUTDOWN_METHOD_NAME` returning `None` actually causes the loop to exit (which I assume is the purpose of that method?) then doing the same thing here should too?

Reply:
No actually. Returning None there returns an empty response to the IDE as per the BSP spec. The IDE then sends an "exit notification" (which does not require a response) to actually trigger the exit of the BSP server process. It was designed this way so that the shutdown method could return an error response if it so chose. If it also exited, then it might cause the response to be lost.

Reply:
As it turns out, rereading the library's source, all I had to do apparently was close the inbound file handle and the read-dispatch loop exits. I tested that it exits cleanly and does not need to set the "kill switch" since the goal code returns cleanly.

Reply:
Got it. And the semantics of `exit` are not "please tear down the server"?

Reply:
The semantics of exit are to exit the process:

> Like the language server protocol, a notification to ask the server to exit its process. The server should exit with success code 0 if the shutdown request has been received before; otherwise with error code 1.

https://build-server-protocol.github.io/docs/specification.html#exit-build-notification

(Although I didn't implement basing the exit code on whether or not a shutdown request was received first.